src/common/types/string_heap.cpp

```diff
@@ -21,7 +21,7 @@ void StringHeap::Move(StringHeap &other) {
 }
 
 string_t StringHeap::AddString(const char *data, idx_t len) {
-	D_ASSERT(Utf8Proc::Analyze(data, len) != UnicodeType::INVALID);
+	D_ASSERT_HEAVY(Utf8Proc::Analyze(data, len) != UnicodeType::INVALID);
```

Comment:
Finally found one heavy-weight assertion that is not optimized out with this PR, because it's a big function that can't be inlined. Our options are using a special macro that _is_ empty in release, or maybe marking such functions as `__attribute__((pure))` or even `((const))`.

Reply:
We don't want to have to actively worry about assertions impacting performance. I can imagine many more of such cases will pop up in the future that will be very hard to debug/profile - not to mention there might be plenty of cases where performance is degraded by 10% that we will not notice. The drawbacks of this seem far larger than any potential benefit we might get from improved optimizations.

Can we swap this around and leave the current version of `D_ASSERT` alone, and create a new `D_UNREACHABLE` (or so) that performs the new variant of the assert? Then we can add that in certain scenarios as needed.

Reply:
Yep, this approach makes sense. If we had `[[assume (expression)]]` from c++23, we wouldn't have to worry about these cases, but it's not there yet. I'll try to investigate the speedups to see which assertions can benefit from the new macro. I'd call it `D_ASSUME`.